### PR TITLE
asm/f64: Updated axpy assembly to wide, pipelined loops.

### PR DESCRIPTION
**Pull request by Kunde21**
Continuing gonum/internal#52 because the floor is lava.

@kortschak I addressed your comments from the review.

### CODE REVIEW

File: asm/f64/benchAxpy_test.go
Comment (kortschak) [inline asm/f64/benchAxpy_test.go]: `const testLen = 1e6`

File: asm/f64/benchAxpy_test.go
Comment (kortschak) [inline asm/f64/benchAxpy_test.go]: These only go to 3e4, but you've allocated 1e6. I'd either test harder, or allocate less (or both).

File: asm/f64/benchAxpy_test.go
Comment (kortschak) [inline asm/f64/benchAxpy_test.go]: a = 2.0

File: asm/f64/asm_test.go
Comment (kortschak) [inline asm/f64/asm_test.go]: Maybe rename this to something like `benchSink` then.

**Review by kortschak — COMMENTED** [review]
I don't understand, a96844f6726a34668d6253cc97e38c222ca26825 has much more than just the register aliasing going on now.

**Comment by kortschak** [comment]
In general, I'd suggest that you avoid remerging master into your work branches, it leads to the kind of problems we saw earlier. I have a workflow that involves a set of targetted commits for a PR and then gets follow-up commits from either me noticing things or PR comments. These follow-ups are labelled carefully so that I know which target they belong to and are then squashed into there target with an interactive rebase. I also alway work from the HEAD of master (rebasing onto master if necessary). This makes things much easier in the long run.

**Comment by Kunde21** [comment]
I'm not sure what's going on with a96844f now.  The diff is showing the newer code as old code.

**Comment by kortschak** [comment]
I know this is horrible, but would you be able to rebase your work onto master so we can work from a clean start. Then `push -f` to your branch.

**Comment by Kunde21** [comment]
There.  That makes more sense, now.

**Comment by kortschak** [comment]
OK. Now, if you do the following, we get back the linear set of commits with no empty merge commits.

```
git checkout f64Axpy
git fetch https://github.com/gonum/internal.git
git rebase FETCH_HEAD
git push -f origin f64Axpy # I am trusting you point origin at your own repo.
```

I just tried the equivalent locally and this works, but if it doesn't, you can always salvage by `git reset --hard 1c227866cea4e5015ebf6930437e8447e78d8536`.

**Comment by kortschak** [comment]
I'll look at c2572119baa80a91424fd959a7fcca5a16577be0 since it looks a little different to the last time I reviewed that set of changes - I could get to that in the next few days. The rest is good. You could do an interactive rebase like this:

```
pick 649fc44 asm/f64: Updated Axpy benchmarks with cleaner go1.7 constructs.
pick 7af3e4d asm/f64: Added and updated testing helper functions.
squash 485e106 asm/f64:  Benchmark updates for PR comments.
squash 3ea382d asm/f64:  Renamed gs placeholder to benchSink.
pick aee6591 asm/f64: Defined macros for named variable registers in axpy* code.
pick c257211 asm/f64:  Cleaned up axpy* assembly code.
```

**Review by kortschak — APPROVED** [review]
LGTM but please wait for another reviewer.

@vladimir-ch can you take a look at this - particularly the last commit.

**Comment by vladimir-ch** [comment]
I'll try.

**Comment by kortschak** [comment]
@Kunde21 because these reviews take such a long time, can we make this the last PR against github.com/gonum/internal and port the remaining four to github.com/gonum/gonum when it becomes open for changes? At the moment, these PRs are what is blocking the merging of repos.

**Comment by Kunde21** [comment]
@kortschak That's not a problem for me.  I'd like to get this one and Scal in before the merge, but if this is the only one, so be it.

I can always work around the merge schedule/needs.

File: asm/f64/benchAxpy_test.go
Comment (vladimir-ch) [inline asm/f64/benchAxpy_test.go:52]: I'm not sure, could `y` overflow and operations on Inf be slower (or faster) than normal?

File: asm/f64/asm_test.go
Comment (vladimir-ch) [inline asm/f64/asm_test.go:107]: Won't this allocate more than necessary?

File: asm/f64/stubs_test.go
Comment (vladimir-ch) [inline asm/f64/stubs_test.go]: Rename to `want` and `x`?

File: asm/f64/stubs_test.go
Comment (vladimir-ch) [inline asm/f64/stubs_test.go]: Also here?

**Review by vladimir-ch — APPROVED** [review]

File: asm/f64/asm_test.go
Comment (Kunde21) [reply asm/f64/asm_test.go:107]: No, this expansion is how we add the increment and guard values in between the test values.  
The resulting slice is:  `[<gdLen*gdVal>,<vec[0]>,<(inc-1)*gdVal>,<vec[1]>,<(inc-1)*gdVal>,...,<gdLen*gdVal>]`.

Comment (vladimir-ch) [reply asm/f64/asm_test.go:107]: Yes, and there are only `len(vec)-1` in-betweens, so it should be enough to have `inrLen := (len(vec)-1)*inc + 1` with care taken when `len(vec) == 0`. However, for testing it doesn't matter if the guard value comes from in-between or from an end-guard, so it's probably ok to leave as is. I'm just noting this to ask if you are aware of this.

File: asm/f64/benchAxpy_test.go
Comment (Kunde21) [reply asm/f64/benchAxpy_test.go:52]: `y` could overflow, but that shouldn't change the operation timing.  The optimization manual only notes value-dependent latency changes for `DIVPD/DIVSD` and `SQRTPD/SQRTSD` instructions.

File: asm/f64/asm_test.go
Comment (Kunde21) [reply asm/f64/asm_test.go:107]: If `gdLen < inc`, going past `n` isn't caught in the tests.  This calculation started the way you're suggesting, until the tail block was unnecessarily running and only the small-increment tests failed.

**Comment by Kunde21** [comment]
I squashed the fixes to clean up the log, but I left the L1Norm fix as a separate commit.  It's ready to merge if that's not an issue.

**Comment by kortschak** [comment]
That all looks good to go. Thanks.

**Pull request merged**
